src/sources/file_descriptors/pipe.rs

```diff
@@ -0,0 +1,191 @@
+use std::fs::File;
+use std::io;
+use std::os::raw::c_int;
+use std::os::unix::io::FromRawFd;
+
+use super::{file_descriptor_source, FileDescriptorConfig};
+use codecs::decoding::{DeserializerConfig, FramingConfig};
+use indoc::indoc;
+use vector_config::configurable_component;
+use vector_core::config::LogNamespace;
+
+use crate::{
+    config::{GenerateConfig, Output, Resource, SourceConfig, SourceContext, SourceDescription},
+    serde::default_decoding,
+    shutdown::ShutdownSignal,
+    SourceSender,
+};
+/// Configuration for the `pipe` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct PipeConfig {
+    /// The maximum buffer size, in bytes, of incoming messages.
+    ///
+    /// Messages larger than this are truncated.
+    #[serde(default = "crate::serde::default_max_length")]
+    pub max_length: usize,
+
+    /// Overrides the name of the log field used to add the current hostname to each event.
+    ///
+    /// The value will be the current hostname for wherever Vector is running.
+    ///
+    /// By default, the [global `host_key` option](https://vector.dev/docs/reference/configuration//global-options#log_schema.host_key) is used.
+    pub host_key: Option<String>,
+
+    #[configurable(derived)]
+    pub framing: Option<FramingConfig>,
+
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    pub decoding: DeserializerConfig,
+
+    /// The file descriptor number to read from.
+    pub fd: c_int,
+}
+
+impl FileDescriptorConfig for PipeConfig {
+    fn host_key(&self) -> Option<String> {
+        self.host_key.clone()
+    }
+    fn framing(&self) -> Option<FramingConfig> {
+        self.framing.clone()
+    }
+    fn decoding(&self) -> DeserializerConfig {
+        self.decoding.clone()
+    }
+    fn description(&self) -> String {
+        format!("fd {}", self.fd)
+    }
+}
+
+inventory::submit! {
+    SourceDescription::new::<PipeConfig>("pipe")
```

Comment:
Not a huge deal but might be nice to make this a const at the top of the file and reference it within like:

```Rust
const NAME: &str = "pipe";
```